presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -231,4 +233,9 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
                 .map(taskContext -> taskContext.accept(visitor, context))
                 .collect(toList());
     }
+
+    public Optional<TaskContext> getTaskContextByTaskId(TaskId taskId)
+    {
+        return taskContexts.stream().filter(context -> context.getTaskId().equals(taskId)).findFirst();
```

Comment:
This will probably not be a perf issue, but it also does't hurt to persist it at `UpdateSystemMemory` (as we already hold onto that reference through `QueryContext`).